pkg/apis/sources/v1alpha1/kuberneteseventsource_types.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/knative/pkg/apis/duck"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Check that KubernetesEventSource can be validated and can be defaulted.
+var _ runtime.Object = (*KubernetesEventSource)(nil)
+
+// Check that KubernetesEventSource implements the Conditions duck type.
+var _ = duck.VerifyType(&KubernetesEventSource{}, &duckv1alpha1.Conditions{})
+
+// KubernetesEventSourceSpec defines the desired state of the source.
+type KubernetesEventSourceSpec struct {
+	// Sink is a reference to an object that will resolve to a domain name to use
+	// as the sink.
+	// +optional
+	Sink *corev1.ObjectReference `json:"sink,omitempty"`
+}
+
+const (
+	// KubernetesEventSourceConditionReady has status True when the
+	// source is ready to send events.
+	KubernetesEventSourceConditionReady = duckv1alpha1.ConditionReady
+
+	// KubernetesEventSourceConditionSinkProvided has status True when the source
+	// has been configured with a sink target.
+	KubernetesEventSourceConditionSinkProvided duckv1alpha1.ConditionType = "SinkProvided"
+)
+
+var kubernetesEventSourceCondSet = duckv1alpha1.NewLivingConditionSet()
+
+// KubernetesEventSourceStatus defines the observed state of the source.
+type KubernetesEventSourceStatus struct {
+	// Conditions holds the state of a source at a point in time.
+	// +optional
+	// +patchMergeKey=type
```

Comment:
What do these do? for ducks we've been using jsonpatch.

Reply:
or, did that change?

Reply:
I think these enable the Patch operation but I'm not sure. They're used in the corev1 apis [here](https://github.com/kubernetes/api/blob/6e4b5aa967ee9a03868e3b4edad98f27e54894c9/core/v1/types.go#L2066) and many other places.